mt-kahypar/datastructures/static_graph.cpp

```diff
@@ -49,6 +52,58 @@ namespace mt_kahypar::ds {
   StaticGraph StaticGraph::contract(parallel::scalable_vector<HypernodeID>& communities, bool /*deterministic*/) {
     ASSERT(communities.size() == _num_nodes);
 
+    // helper function that is used in multiple places for edge deduplication
+    auto deduplicate_tmp_edges = [](TmpEdgeInformation* edge_start, TmpEdgeInformation* edge_end) {
+      ASSERT(std::distance(edge_start, edge_end) >= 0);
+      std::sort(edge_start, edge_end,
+                [](const TmpEdgeInformation& e1, const TmpEdgeInformation& e2) {
+                  return e1._target < e2._target;
+                });
+
+      // Deduplicate, aggregate weights and calculate minimum unique id
+      //
+      // <-- deduplicated --> <-- already processed --> <-- to be processed --> <-- invalid edges -->
+      //                    ^                         ^
+      // valid_edge_index ---        tmp_edge_index ---
+      size_t valid_edge_index = 0;
+      size_t tmp_edge_index = 1;
+      while (tmp_edge_index < static_cast<size_t>(std::distance(edge_start, edge_end)) && edge_start[tmp_edge_index].isValid()) {
+        HEAVY_COARSENING_ASSERT(
+          [&](){
+            size_t i = 0;
+            for (; i <= valid_edge_index; ++i) {
+              if (!edge_start[i].isValid()) {
+                return false;
+              } else if ((i + 1 <= valid_edge_index) &&
+                edge_start[i].getTarget() >= edge_start[i + 1].getTarget()) {
+                return false;
+              }
+            }
+            for (; i < tmp_edge_index; ++i) {
+              if (edge_start[i].isValid()) {
+                return false;
+              }
+            }
+            return true;
+          }(),
+          "Invariant violated while deduplicating incident edges!"
+        );
+
+        TmpEdgeInformation& valid_edge = edge_start[valid_edge_index];
+        TmpEdgeInformation& next_edge = edge_start[tmp_edge_index];
+        if (valid_edge.getTarget() == next_edge.getTarget()) {
+          valid_edge.addWeight(next_edge.getWeight());
+          valid_edge.updateID(next_edge.getID());
```

Comment:
Is this step necessary if we reindex the edges later on anyways?

Reply:
The purpose of this is to get the same id for the edge itself and its reverse edge. I don't know how this should work otherwise

Reply:
This is only a minor suggestion. Feel free to ignore it, if it's too much work. My point is that we rebuild the edge IDs later -- to get a compact range of IDs. We're not using them in the meantime, other than for the rebuild itself, which should be possible to do without.

Reply:
If you can describe a algorithm for this, sure. I've thought a while about this when implementing it and I don't have an idea for an algorithm that can construct the IDs afterwards efficiently (at least not without also significantly changing the construction of the edges themselves)